File: khipu-studio/khipu-web/src/lib/projects.ts
import api from './api';

export interface Project {
  id: string;
  tenant_id: string;
  owner_id: string;
  title: string;
  subtitle?: string;
  authors?: string[];
  narrators?: string[];
  translators?: string[];
  adaptors?: string[];
  language: string;
  description?: string;
  publisher?: string;
  publish_date?: string;
  isbn?: string;
  cover_image_url?: string;
  status: 'draft' | 'in_progress' | 'review' | 'completed' | 'published' | 'archived';
  settings?: Record<string, unknown>;
  workflow_completed?: Record<string, boolean>;
  created_at: string;
  updated_at: string;
  completed_at?: string;
  archived_at?: string;
}

export interface ProjectCreate {
  title: string;
  subtitle?: string;
  authors?: string[];
  narrators?: string[];
  language?: string;
  description?: string;
}

export interface ProjectUpdate {
  title?: string;
  subtitle?: string;
  authors?: string[];
  narrators?: string[];
  translators?: string[];
  adaptors?: string[];
  language?: string;
  description?: string;
  publisher?: string;
  publish_date?: string;
  isbn?: string;
  status?: 'draft' | 'in_progress' | 'review' | 'completed' | 'published' | 'archived';
  settings?: Record<string, unknown>;
  workflow_completed?: Record<string, boolean>;
  cover_image_url?: string;
}

export interface ProjectsListResponse {
  items: Project[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export const projectsApi = {
  list: async (params?: {
    skip?: number;
    limit?: number;
    search?: string;
    status?: string;
  }): Promise<ProjectsListResponse> => {
    const response = await api.get('/projects/', { params });
    return response.data;
  },

  get: async (id: string): Promise<Project> => {
    const response = await api.get(`/projects/${id}`);
    return response.data;
  },

  create: async (data: ProjectCreate): Promise<Project> => {
    const response = await api.post('/projects/', data);
    return response.data;
  },

  update: async (id: string, data: ProjectUpdate): Promise<Project> => {
    const response = await api.put(`/projects/${id}`, data);
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/projects/${id}`);
  },

  suggestIpa: async (id: string, word: string): Promise<{
    success: boolean;
    ipa?: string;
    error?: string;
    examples?: string[];
    source?: string;
  }> => {
    const response = await api.post(`/projects/${id}/suggest-ipa`, { word });
    return response.data;
  },
};

File: khipu-studio/khipu-cloud-api/shared/schemas/projects.py
"""Project schemas."""
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, field_validator


class ProjectBase(BaseModel):
    """Base project schema."""
    title: str = Field(..., min_length=1, max_length=500)
    subtitle: Optional[str] = Field(None, max_length=500)
    authors: Optional[List[str]] = None
    narrators: Optional[List[str]] = None
    translators: Optional[List[str]] = None
    adaptors: Optional[List[str]] = None
    language: str = Field(default="es-PE", max_length=10)
    description: Optional[str] = None
    publisher: Optional[str] = Field(None, max_length=255)
    publish_date: Optional[datetime] = None
    isbn: Optional[str] = Field(None, max_length=20)


class ProjectCreate(ProjectBase):
    """Schema for creating a project."""
    pass


class ProjectUpdate(BaseModel):
    """Schema for updating a project."""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    subtitle: Optional[str] = Field(None, max_length=500)
    authors: Optional[List[str]] = None
    narrators: Optional[List[str]] = None
    translators: Optional[List[str]] = None
    adaptors: Optional[List[str]] = None
    language: Optional[str] = Field(None, max_length=10)
    description: Optional[str] = None
    publisher: Optional[str] = Field(None, max_length=255)
    publish_date: Optional[datetime] = None
    isbn: Optional[str] = Field(None, max_length=20)
    status: Optional[str] = None
    settings: Optional[Dict[str, Any]] = None
    workflow_completed: Optional[Dict[str, Any]] = None
    
    # Note: Blob storage settings are stored in the settings JSON field with structure:
    # settings.creds.storage.azure = { accountName, accessKey, containerName, endpoint }


class ProjectResponse(ProjectBase):
    """Schema for project response."""
    id: str
    tenant_id: str
    owner_id: str
    status: str
    cover_image_url: Optional[str] = None
    manuscript_word_count: Optional[int] = None
    manuscript_character_count: Optional[int] = None
    workflow_completed: Dict[str, Any] = {}
    settings: Dict[str, Any] = {}
    created_at: datetime
    updated_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    archived_at: Optional[datetime] = None

    @field_validator('id', 'tenant_id', 'owner_id', mode='before')
    @classmethod
    def convert_uuid_to_str(cls, v: Any) -> str:
        """Convert UUID to string."""
        return str(v)

    class Config:
        from_attributes = True


class ProjectListResponse(BaseModel):
    """Schema for project list response."""
    items: List[ProjectResponse]
    total: int
    page: int
    page_size: int
    pages: int
